File: ruby/lib/xstreamly.rb
require 'rubygems'
require 'json' #gem install json
require 'base64'
require 'cgi'
require 'openssl'
require 'net/https'
require 'net/http'

module XStreamly
	class Client
    def initialize(appKey,email,password)
      @appKey = appKey
      @email = email
      @password = password
      @http = Net::HTTP.new('api.x-stream.ly', 443)
      @http.use_ssl = true
    end
		
    def send(channel, eventName, data,persisted = false)
	    req = Net::HTTP::Post.new(URI.encode('/api/v1.1/'+@appKey+'/channels/'+channel+'/events/'+eventName+'?persisted='+persisted.to_s), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password

	    req.body = data
	    response = @http.request(req)
	    result = response.code.to_i
      case result
          when 200..299
          	return true
          else 
          	raise RuntimeError, "Unknown error (status code #{result}): #{response.body}"
          end
    end
        
    def setCallback(channel,endPoint,secret,eventName)
      req = Net::HTTP::Post.new(URI.encode('/api/v1.1/'+@appKey+'/feeds/out/custom'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password

	    req.body = '{"channel":"'+channel+'","endpoint":"'+endPoint+'", "secret":"'+secret+'", "event": "'+eventName+'"}'
	    response = @http.request(req)
          
      case response.code.to_i
        when 200..299
        	return response.body
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
        
    def removeCallback(index)
      req = Net::HTTP::Delete.new(URI.encode('/api/v1.1/'+@appKey+'/feeds/out/custom/'+index.to_s), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
      response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return true
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
        
    def getCallbacks()
    	req = Net::HTTP::Get.new(URI.encode('/api/v1.1/'+@appKey+'/feeds/out/custom'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return (JSON(response.body))['items']
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def setTwitterStream(channel,event,requestData)
      req = Net::HTTP::Post.new(URI.encode('/api/v1.1/'+@appKey+'/feeds/in/twitter'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password

	    req.body = '{"channel":"'+channel+'","event":"'+event+'", "requestData":"'+requestData+'"}'
	    response = @http.request(req)
          
      case response.code.to_i
        when 200..299
        	return response.body
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def generateRequestData(trackParameter,consumerKey,consumerSecret,accessToken,accessTokenSecret)
      #generate a random nonce
      nonce  =(0...10).map{65.+(rand(25)).chr}.join
      timestamp = (Time.now.to_i).to_s
      track = CGI.escape(trackParameter)
      
      base = "POST&https%3A%2F%2Fstream.twitter.com%2F1%2Fstatuses%2Ffilter.json&" +
        "oauth_consumer_key%3D" + consumerKey +
        "%26oauth_nonce%3D"+nonce +
        "%26oauth_signature_method%3DHMAC-SHA1%26oauth_timestamp%3D" + timestamp +
        "%26oauth_token%3D" + accessToken +
        "%26oauth_version%3D1.0"+ 
        "%26track%3D" +CGI.escape(track)
        
      key = consumerSecret +'&'+CGI.escape(accessTokenSecret);
      
      digest  = OpenSSL::Digest::Digest.new('sha1')
      signature=CGI.escape(Base64.encode64(OpenSSL::HMAC.digest(digest,key,base)).strip)
      
      requestData = "oauth_signature=" + signature +
        "&oauth_token=" + accessToken +
        "&track=" + track +
        "&oauth_consumer_key=" + consumerKey +
        "&oauth_signature_method=HMAC-SHA1&oauth_version=1.0" +
        "&oauth_nonce=" + nonce +
        "&oauth_timestamp=" + timestamp
      
      return requestData
    end
        
    def removeTwitterStream(index)
      req = Net::HTTP::Delete.new(URI.encode('/api/v1.1/'+@appKey+'/feeds/in/twitter/'+index.to_s), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
      response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return true
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
        
    def getTwitterStreams()
    	req = Net::HTTP::Get.new(URI.encode('/api/v1.1/'+@appKey+'/feeds/in/twitter'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return (JSON(response.body))['items']
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def activeChannels()
      req = Net::HTTP::Get.new(URI.encode('/api/v1.1/'+@appKey+'/activeChannels'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return (JSON(response.body))['channels']
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def usageConnections()
      req = Net::HTTP::Get.new(URI.encode('/usage/connections'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return (JSON(response.body))['data']
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def usageMessages()
      req = Net::HTTP::Get.new(URI.encode('/usage/messages'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return (JSON(response.body))['data']
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def getTokens()
      req = Net::HTTP::Get.new(URI.encode('/api/v1.1/'+@appKey+'/security'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return (JSON(response.body))['sessions']
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end
    
    def createToken(read=true,write=true,channel=nil,event=nil,source=nil,isPrivate=false)
      req = Net::HTTP::Post.new(URI.encode('/api/v1.1/'+@appKey+'/security'), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    data = {}
	    if(read && write)
	      # don't set the value then
      elsif (read)
        data['action']='read|view'
      elsif (write)
        data['action']='write|view'
      else
        data['action']='view'
      end
      
      if(!channel.nil?)
        data['channel']=channel
      end
      
      if(!event.nil?)
        data['event']=event
      end
      
      if(!source.nil?)
        data['source']=source
      end
      
      data['private']=isPrivate
	    
	    req.body = data.to_json
	    
	    puts req.body
	    
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return response.body
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end

    def deleteToken(token)
      req = Net::HTTP::Delete.new(URI.encode('/api/v1.1/'+@appKey+'/security/'+token), initheader = {'Content-Type' =>'application/json'})
	    req.basic_auth @email, @password
	    response = @http.request(req)
	
      case response.code.to_i
        when 200..299
        	return true
        else 
        	raise RuntimeError, "Unknown error (status code #{response.code} ): #{response.body}"
    	end
    end

  end
end

File: php/src/xstreamly.php
<?php

namespace XStreamly;

if(!function_exists('curl_init')){
  throw new \Exception('XStreamly need the CURL PHP extension.');
}

if(!function_exists('json_decode')){
  throw new \Exception('XStreamly need the JSON PHP extension.');
}

class XStreamlyClient{
  private $appKey;
  private $email;
  private $password;
  private $baseUrl = 'https://secure.x-stream.ly'; 

  public function __construct($appKey,$email,$password){
    $this->appKey = $appKey;
    $this->email = $email;
    $this->password = $password;
  }
  
  public function send($channel, $eventName, $data,$persisted = false){
    return $this->genericPost('/api/v1.1/' . $this->appKey . '/channels/' . $channel . '/events/' . $eventName . '?persisted=' . ($persisted?'true':'false'),$data);
  }
  
  //should be an array of objects with properties channel,event,message,[persisted]
  public function sendBatch($messages){
    return $this->genericPost('/api/v1.1/' . $this->appKey . '/messages',$messages);
  }
  
  public function setCallback($channel, $endPoint,$secret, $eventName){
    $data = array('channel'=>$channel,'endpoint'=>$endPoint,'secret'=>$secret,'event'=>$eventName);
    return $this->genericPost('/api/v1.1/' . $this->appKey . '/feeds/out/custom',$data);
  }
  
  public function removeCallback($key){
    return $this->genericDelete('/api/v1.1/' . $this->appKey . '/feeds/out/custom/'.$key);
  }
  
  public function getCallbacks(){
    return $this->genericGet('/api/v1.1/' . $this->appKey . '/feeds/out/custom')->items;
  }
  
  public function getActiveChannels(){
    return $this->genericGet('/api/v1.1/' . $this->appKey . '/activeChannels');
  }
  
  public function getConnectionUsage(){
    return $this->genericGet('/usage/connections');
  }
  
  public function getMessageUsage(){
    return $this->genericGet('/usage/messages');
  }
  
  public function getTokens($channel=null, $eventName=null, $source=null){
    $params = '';
    
    if(!is_null($channel)){
      $params .='channel='.$channel;
    }
    
    if(!is_null($eventName)){
      $params .='event='.$eventName;
    }
    
    if(!is_null($source)){
      $params .='source='.$source;
    }
    
    if(strlen($params)){
      $params = "?" . $params;
    }
  
    return $this->genericGet('/api/v1.1/' . $this->appKey . '/security/' . $params);
  }
  
  public function createToken($canRead = true, $canWrite = true, $channel = null, $event = null, $source = null, $isPrivate = false){
    $data = array();
    
    if(!$canRead || !$canWrite){
      if($canWrite){
        $action = 'read';
      } else {
        $action = 'write';
      }
      
      $data['action']=$action;
    }
    
    
    if(null!=$channel){
      $data['channel']=$channel;
    }
    
    if(null!=$event){
      $data['event']=$event;
    }
    
    if(null!=$source){
      $data['source']=$source;
    }
    
    return $this->genericPost('/api/v1.1/' . $this->appKey . '/security',$data);
  }
  
  public function deleteToken($token){
    return $this->genericDelete('/api/v1.1/' . $this->appKey . '/security/' . $token);
  }
  
  private function genericDelete($url){
    $request = $this->getRequest($url);
    curl_setopt($request,CURLOPT_CUSTOMREQUEST,'DELETE');
    
    return $this->execute($request);
  }
  
  private function genericGet($url){
    $request = $this->getRequest($url);
        
    return json_decode($this->execute($request));
  }
  
  private function genericPost($url,$data){
    $request = $this->getRequest($url);
    
    $stringData =json_encode($data);
        
    curl_setopt($request,CURLOPT_POST,1);
    curl_setopt($request,CURLOPT_HTTPHEADER,array('Content-type: application/json'));
    curl_setopt($request,CURLOPT_POSTFIELDS,$stringData);
    
    return $this->execute($request);   
  }
  
  private function getRequest($url){
    $request = curl_init($this->baseUrl . $url);
    curl_setopt($request,CURLOPT_USERPWD,$this->email.':'.$this->password);
    curl_setopt($request,CURLOPT_RETURNTRANSFER,1);
    curl_setopt($request,CURLOPT_FAILONERROR,1);
    return $request;
  }
  
  private function execute($request){
    $response = curl_exec($request);
       
    $error = curl_error($request);
    if($error){
      throw new \Exception("couldn't send message, got error: ". $error);
    }
    
    $http_status = curl_getinfo($request,CURLINFO_HTTP_CODE);
    curl_close($request);
    
    if($http_status ==302 && strpos($response,'login.html')){
      throw new \Exception('Sorry your credentials are invalid, please check your app Key, e-mail address and password to make sure they are correct.');
    }
    
    if($http_status == 200 || $http_status == 202){
      return $response;
    } else {
      throw new \Exception("couldn't send message, got response code: ". $http_status. " response: ".$response);
    }
  }
}

?>
